Library/Homebrew/livecheck/strategy/pypi.rb

```diff
@@ -1,19 +1,21 @@
 # typed: strict
 # frozen_string_literal: true
 
+require "json"
+require "utils/curl"
+
 module Homebrew
   module Livecheck
     module Strategy
       # The {Pypi} strategy identifies versions of software at pypi.org by
-      # checking project pages for archive files.
+      # using the JSON API endpoint.
       #
-      # PyPI URLs have a standard format but the hexadecimal text between
-      # `/packages/` and the filename varies:
+      # PyPI URLs have a standard format:
       #
       # * `https://files.pythonhosted.org/packages/<hex>/<hex>/<long_hex>/example-1.2.3.tar.gz`
       #
-      # As such, the default regex only targets the filename at the end of the
-      # URL.
+      # This method uses the `info.version` field in the JSON response to
+      # determine the latest stable version.
```

Comment:
One thing worth confirming is whether the `version` returned from the API is simply the newest release or if it's the newest non-yanked release. The `info` object in the JSON field also has a `yanked` boolean field (like the objects in the `releases` and `urls` arrays), which implies that it may be possible for the newest version to be a yanked release.

To confirm this, we would need to find a package where the latest release is a yanked release to see what the API returns. If the `version` value isn't guaranteed to be a non-yanked version, then we would need to iterate through `urls` (`releases` is deprecated, according to the [JSON API documentation](https://docs.pypi.org/api/json/)).

Reply:
Might need to find an example to test, but the code may prioritize non-yanked/non-prerelease:
* https://github.com/pypi/warehouse/blob/main/warehouse/routes.py#L618-L619
  ```py
          "/pypi/{name}/json",
          factory="warehouse.legacy.api.json.latest_release_factory",
  ```
* https://github.com/pypi/warehouse/blob/main/warehouse/legacy/api/json.py#L211-L214
  ```py
              .order_by(
                  Release.yanked.asc(),
                  Release.is_prerelease.nullslast(),
                  Release._pypi_ordering.desc(),
  ```

In which case, the `yanked` boolean field would only be `true` if all releases are yanked.

Reply:
Oh nice, thanks for digging up related code. I was hoping that they prioritized the latest non-yanked release. It still wouldn't hurt to confirm with a package where the newest release is yanked but that's reassuring enough for us to move ahead with this PR for now.

I wonder if there's an intelligent way of easily finding a PyPI project where the newest release is yanked besides checking random packages until we find a match (that would be like finding a needle in a haystack) 😆

It would probably be easier to create a PyPI project for testing, create some test releases, yank the newest one, check the API response, and then delete the project. I don't know anything about PyPI, so I'm not sure how okay they would be with that but that may be an idea.

Reply:
Did find one from searching Gentoo's commit history. 

https://pypi.org/project/cleo/#history
```console
❯ curl -sL https://pypi.org/pypi/cleo/json | jq -r '.info | [.version, .yanked] | @csv'
"2.1.0",false
```

Reply:
Oh wow, great work finding that example. That's exactly how I was hoping PyPI would handle that situation (i.e., 2.2.1 and and 2.2.0 are the newest releases and they're both yanked, so PyPI uses 2.1.0 as the `version` in the JSON API since it's the newest non-yanked release). Thanks!